src/Components/Components/src/Routing/TemplateSegment.cs

```diff
@@ -25,6 +39,10 @@ public TemplateSegment(string template, string segment, bool isParameter)
                     throw new ArgumentException($"Malformed parameter '{segment}' in route '{template}' has no name before the constraints list.");
                 }
 
+                // Set the IsOptional flag to true if any type constraints
+                // for this parameter are designated as optional.
+                IsOptional = tokens.Skip(1).Any(token => token.Contains('?'));
+
                 Value = tokens[0];
                 Constraints = tokens.Skip(1)
                     .Select(token => RouteConstraint.Parse(template, segment, token))
```

Comment:
The constraint itself wouldn't be treated as optional, but the segment would be. The result of this is as follows:

- If the user has a route like `/a/{id:int:long?}/literal`, then it will through an error because they have provided a non-optional segment after an optional one. It makes sense to me to throw an error in this case but let me know if you think otherwise.
- When matching it is not required that a match be found for this segment since it contains an optional component.

You'd still get two separate constraints to match with, one optional and one not.